repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -238,54 +240,53 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
       }
     }
 
+    def buildTenantToRolesMap(token: ValidToken): TenantToRolesMap = {
+      token.defaultTenantId.map(_ -> Set.empty[String]).toMap ++
+        token.roles.groupBy(_.tenantId).map({ case (key, value) => key.getOrElse(DomainRoleTenantKey) -> value.map(_.name).toSet })
+    }
+
     def buildTenantHeader(defaultTenant: Option[String],
-                          roleTenants: Seq[String],
-                          matchedUriTenant: Option[String]): Vector[String] = {
+                          roleTenants: Set[String],
+                          matchedTenants: Set[String]): Set[ValueWithQuality] = {
       val sendAllTenants = configuration.getTenantHandling.isSendAllTenantIds
       val sendTenantIdQuality = Option(configuration.getTenantHandling.getSendTenantIdQuality)
       val sendQuality = sendTenantIdQuality.isDefined
       val defaultTenantQuality = sendTenantIdQuality.map(_.getDefaultTenantQuality).getOrElse(0.0)
-      val uriTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
+      val matchedTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
       val rolesTenantQuality = sendTenantIdQuality.map(_.getRolesTenantQuality).getOrElse(0.0)
 
-      case class PreferredTenant(id: String, quality: Double)
-
-      val preferredTenant = (defaultTenant, matchedUriTenant) match {
-        case (Some(default), Some(uri)) =>
-          val quality = if (default.equals(uri)) {
-            math.max(defaultTenantQuality, uriTenantQuality)
-          } else {
-            uriTenantQuality
+      val preferredTenants =
+        if (matchedTenants.nonEmpty) {
+          matchedTenants map {
+            case matchedTenant if defaultTenant.exists(matchedTenant.equals) =>
```

Comment:
This is the carryover of the default tenant quality override. I just wanted to comment to say that I don't love that we give special treatment to this case. It's not like we always use the highest applicable quality, just when the default tenant ID matches. This leads to more complex behavior which may not always be obvious to a user.

Reply:
Man, I thought we were using the highest quality that matched. You should flag this behavior for removal in 9.x. It's possible this behavior is needed by someone though. It's also possible it's just a hold over from a different world.

Reply:
Will do.